File: userspace/coreutils-rs/src/wc.rs
//! wc — count lines, words, and bytes.
#![no_std]
#![no_main]

use syscall_lib::{
    O_RDONLY, STDERR_FILENO, STDOUT_FILENO, close, open, read, write_str, write_u64,
};

syscall_lib::entry_point!(main);

fn wc_fd(fd: i32) -> (u64, u64, u64) {
    let mut buf = [0u8; 4096];
    let mut lines: u64 = 0;
    let mut words: u64 = 0;
    let mut bytes: u64 = 0;
    let mut in_word = false;
    loop {
        let n = read(fd, &mut buf);
        if n <= 0 {
            break;
        }
        let n = n as usize;
        bytes += n as u64;
        for &b in &buf[..n] {
            if b == b'\n' {
                lines += 1;
            }
            if b == b' ' || b == b'\t' || b == b'\n' || b == b'\r' {
                in_word = false;
            } else if !in_word {
                in_word = true;
                words += 1;
            }
        }
    }
    (lines, words, bytes)
}

fn main(args: &[&str]) -> i32 {
    let mut show_lines = false;
    let mut show_words = false;
    let mut show_bytes = false;
    let mut first_file = 1;

    // Parse flags.
    for (i, arg) in args.iter().enumerate().skip(1) {
        if arg.starts_with('-') && arg.len() > 1 {
            for b in arg.as_bytes()[1..].iter() {
                match b {
                    b'l' => show_lines = true,
                    b'w' => show_words = true,
                    b'c' => show_bytes = true,
                    _ => {}
                }
            }
            first_file = i + 1;
        } else {
            break;
        }
    }

    if !show_lines && !show_words && !show_bytes {
        show_lines = true;
        show_words = true;
        show_bytes = true;
    }

    let files = &args[first_file..];
    let mut total_l: u64 = 0;
    let mut total_w: u64 = 0;
    let mut total_c: u64 = 0;

    if files.is_empty() {
        let (l, w, c) = wc_fd(0);
        if show_lines {
            write_u64(STDOUT_FILENO, l);
            write_str(STDOUT_FILENO, " ");
        }
        if show_words {
            write_u64(STDOUT_FILENO, w);
            write_str(STDOUT_FILENO, " ");
        }
        if show_bytes {
            write_u64(STDOUT_FILENO, c);
        }
        write_str(STDOUT_FILENO, "\n");
        return 0;
    }

    for file in files {
        let bytes = file.as_bytes();
        if bytes.len() > 254 {
            write_str(STDERR_FILENO, "wc: path too long\n");
            continue;
        }
        let mut path = [0u8; 256];
        path[..bytes.len()].copy_from_slice(bytes);
        path[bytes.len()] = 0;

        let fd = open(&path[..=bytes.len()], O_RDONLY, 0);
        if fd < 0 {
            write_str(STDERR_FILENO, "wc: cannot open: ");
            write_str(STDERR_FILENO, file);
            write_str(STDERR_FILENO, "\n");
            continue;
        }
        let (l, w, c) = wc_fd(fd as i32);
        close(fd as i32);
        total_l += l;
        total_w += w;
        total_c += c;

        if show_lines {
            write_u64(STDOUT_FILENO, l);
            write_str(STDOUT_FILENO, " ");
        }
        if show_words {
            write_u64(STDOUT_FILENO, w);
            write_str(STDOUT_FILENO, " ");
        }
        if show_bytes {
            write_u64(STDOUT_FILENO, c);
            write_str(STDOUT_FILENO, " ");
        }
        write_str(STDOUT_FILENO, file);
        write_str(STDOUT_FILENO, "\n");
    }

    if files.len() > 1 {
        if show_lines {
            write_u64(STDOUT_FILENO, total_l);
            write_str(STDOUT_FILENO, " ");
        }
        if show_words {
            write_u64(STDOUT_FILENO, total_w);
            write_str(STDOUT_FILENO, " ");
        }
        if show_bytes {
            write_u64(STDOUT_FILENO, total_c);
            write_str(STDOUT_FILENO, " ");
        }
        write_str(STDOUT_FILENO, "total\n");
    }
    0
}

#[panic_handler]
fn panic(_: &core::panic::PanicInfo) -> ! {
    syscall_lib::exit(101)
}

File: userspace/coreutils/wc.c
/* wc — count lines, words, and bytes */
#include <unistd.h>
#include <fcntl.h>
#include <string.h>

static void write_str(int fd, const char *s) {
    const char *p = s;
    while (*p) p++;
    write(fd, s, p - s);
}

static void write_num(int fd, long long n) {
    char buf[24];
    int i = sizeof(buf) - 1;
    buf[i] = '\0';
    if (n == 0) { buf[--i] = '0'; }
    while (n > 0) { buf[--i] = '0' + (n % 10); n /= 10; }
    write_str(fd, &buf[i]);
}

static int wc_fd(int fd, long long *tl, long long *tw, long long *tc) {
    char buf[4096];
    long long lines = 0, words = 0, bytes = 0;
    int in_word = 0;
    ssize_t n;
    while ((n = read(fd, buf, sizeof(buf))) > 0) {
        bytes += n;
        for (ssize_t i = 0; i < n; i++) {
            if (buf[i] == '\n') lines++;
            if (buf[i] == ' ' || buf[i] == '\t' || buf[i] == '\n' || buf[i] == '\r') {
                in_word = 0;
            } else if (!in_word) {
                in_word = 1;
                words++;
            }
        }
    }
    if (n < 0) write_str(2, "wc: read error\n");
    *tl += lines; *tw += words; *tc += bytes;
    return n < 0 ? -1 : 0;
}

int main(int argc, char **argv) {
    int show_lines = 0, show_words = 0, show_bytes = 0;
    int first_file = 1;

    /* Parse flags. */
    for (int i = 1; i < argc; i++) {
        if (argv[i][0] == '-' && argv[i][1] != '\0') {
            for (const char *p = argv[i] + 1; *p; p++) {
                if (*p == 'l') show_lines = 1;
                else if (*p == 'w') show_words = 1;
                else if (*p == 'c') show_bytes = 1;
            }
            first_file = i + 1;
        } else {
            break;
        }
    }

    /* Default: show all. */
    if (!show_lines && !show_words && !show_bytes) {
        show_lines = show_words = show_bytes = 1;
    }

    long long total_l = 0, total_w = 0, total_c = 0;
    int file_count = argc - first_file;

    int ret = 0;

    if (file_count == 0) {
        /* Read from stdin. */
        if (wc_fd(0, &total_l, &total_w, &total_c) < 0) ret = 1;
        if (show_lines) { write_num(1, total_l); write_str(1, " "); }
        if (show_words) { write_num(1, total_w); write_str(1, " "); }
        if (show_bytes) { write_num(1, total_c); }
        write_str(1, "\n");
        return ret;
    }

    for (int i = first_file; i < argc; i++) {
        int fd = open(argv[i], O_RDONLY);
        if (fd < 0) {
            write_str(2, "wc: cannot open: ");
            write_str(2, argv[i]);
            write_str(2, "\n");
            ret = 1;
            continue;
        }
        long long l = 0, w = 0, c = 0;
        if (wc_fd(fd, &l, &w, &c) < 0) ret = 1;
        close(fd);
        total_l += l; total_w += w; total_c += c;
        if (show_lines) { write_num(1, l); write_str(1, " "); }
        if (show_words) { write_num(1, w); write_str(1, " "); }
        if (show_bytes) { write_num(1, c); write_str(1, " "); }
        write_str(1, argv[i]);
        write_str(1, "\n");
    }

    if (file_count > 1) {
        if (show_lines) { write_num(1, total_l); write_str(1, " "); }
        if (show_words) { write_num(1, total_w); write_str(1, " "); }
        if (show_bytes) { write_num(1, total_c); write_str(1, " "); }
        write_str(1, "total\n");
    }
    return ret;
}
